### PR TITLE
ARROW-4558: [C++][Flight] Implement gRPC customizations without UB

### PR DESCRIPTION
**Pull request by wesm**
I admit this feels gross, but it's less gross than what was there before. I can do some more clean up but wanted to get feedback before spending any more time on it

So, the problem partially lies with the gRPC C++ library. The obvious thing, and first thing I tried, was to specialize `SerializationTraits<protocol::FlightData>` and do casts between `FlightData` and `protocol::FlightData` (the proto) at the last possible moment. Unfortunately, this seems to not be possible because of this:

https://github.com/grpc/grpc/blob/master/include/grpcpp/impl/codegen/proto_utils.h#L100

So I had to override that Googly hack and go to some shenanigans (see protocol.h/protocol.cc) to make sure the same templates are always visible both in `Flight.grpc.pb.cc` as well as our client.cc/server.cc

### CODE REVIEW

**Comment by wesm** [comment]
cc @pitrou @lihalite

**Comment by pitrou** [comment]
Perhaps it would be useful to ping the gRPC community to know if this is idiomatic.

File: cpp/src/arrow/flight/client.cc [hunk 1]
```diff
@@ -127,12 +121,9 @@ class FlightPutWriter::FlightPutWriterImpl : public ipc::RecordBatchWriter {
...
     IpcPayload payload;
     RETURN_NOT_OK(ipc::internal::GetRecordBatchPayload(batch, pool_, &payload));
-    // XXX this cast is undefined behavior
-    auto custom_writer = reinterpret_cast<grpc::ClientWriter<IpcPayload>*>(writer_.get());
-    // Explicitly specify the override to invoke - otherwise compiler
-    // may invoke through vtable (not updated by reinterpret_cast)
-    if (!custom_writer->grpc::ClientWriter<IpcPayload>::Write(payload,
-                                                              grpc::WriteOptions())) {
+
+    if (!writer_->Write(*reinterpret_cast<const pb::FlightData*>(&payload),
```
Comment (pitrou) [inline cpp/src/arrow/flight/client.cc]: Basically we are praying that `Write` doesn't do anything with the FlightData pointer except pass it to `SerializationTraits<FlightData>`, right?

File: cpp/src/arrow/flight/protocol.cc [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+
+#include "arrow/flight/protocol.h"
+
+#include "arrow/flight/Flight.grpc.pb.cc"  // NOLINT
```
Comment (lihalite) [inline cpp/src/arrow/flight/protocol.cc]: So the trick of `protocol.cc` and `protocol.h` is to make sure you can only get the generated Protobuf code along with our template specialization?

**Review by ghost — COMMENTED** [review]
This feels icky, but less icky than before. 

https://github.com/grpc/grpc/issues/15764 is the upstream issue we need, I think (it'd be ideal if whatever API they come up with eventually just gives us a byte stream/sink).

**Comment by pitrou** [comment]
@lihalite Interesting.
Apparently, https://github.com/tensorflow/tensorflow/blob/ced4fddc0400fedce8ba15fb9a3e6e765ddda757/tensorflow/core/distributed_runtime/rpc/grpc_tensor_coding.cc#L171 shows how to do actual zero-copy writes, i.e. write directy from application data to socket - by creating a bunch of slices that point to application data and chaining them in the `ByteBuffer` (which can contain any number of slices).

**Comment by lihalite** [comment]
@pitrou That is awesome, maybe we should do that in a follow-up.

File: cpp/src/arrow/flight/protocol.cc (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/flight/protocol.cc]: Right

File: cpp/src/arrow/flight/client.cc (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/flight/client.cc]: That's right. But the reader/writer layers have no knowledge of the message types; SerializationTraits is the only point of contact

* https://github.com/grpc/grpc/blob/f41affc9c7e0358612b9747e914e8b01af53f75d/include/grpcpp/impl/codegen/call_op_set.h#L395

* https://github.com/grpc/grpc/blob/46bd2f7adb926053345665d5c487fa20acd2b5b0/include/grpcpp/impl/codegen/server_interface.h#L255

**Comment by wesm** [comment]
@pitrou I was going to propose that on the e-mail thread but you found it already =) Composing a slice from smaller zero-copy constituent slices is the way to avoid memcpy there. The only annoyance is that we'll have to add padding slices but c'est la vie

File: cpp/src/arrow/flight/protocol.cc (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/flight/protocol.cc]: It appears that the linker (at least with gcc on Linux) was seeing the vtable generated in Flight.grpc.pb.cc and overriding the ones in server.cc/client.cc. This was the only way I could make sure that the same code is generated in all places

**Comment by lihalite** [comment]
FWIW, the Java side also adds padding slices (it was one of the incompatibilities between the two implementations) 

https://github.com/apache/arrow/blob/27ba26cad39d92713e17d811a65ad0b87c587610/java/flight/src/main/java/org/apache/arrow/flight/ArrowMessage.java#L276-L281

**Comment by pitrou** [comment]
Hmm... why do we need padding slices?

File: cpp/src/arrow/flight/protocol.cc (hunk 2 above)
Comment (lihalite) [reply cpp/src/arrow/flight/protocol.cc]: I think I'd appreciate a comment about that just so that someone else looking at it isn't confused.

**Comment by wesm** [comment]
I commented on the gRPC thread

**Comment by wesm** [comment]
@pitrou a buffer's `size` may not be a multiple of 8. But a buffer should be padded already so we should be able to round up the buffer size to the next multiple of 8 and avoid the padding slices

https://github.com/apache/arrow/blob/master/cpp/src/arrow/flight/serialization-internal.h#L302

**Comment by pitrou** [comment]
Right... but the "padding" won't be necessarily zero-filled (for example if taking a slice of a buffer). Is it a problem?

**Comment by pitrou** [comment]
I looked at Tensorflow's use of gRPC and it seems they don't do zero-copy on the receiving side. For example with the `SendTensor` method in `eager_service.proto`, it seems the server will use regular deserialization for the tensor bytes data sent by the client.

Also there are weird hacks where protobuf-generated definitions are replaced/overriden with a hand-written ones, I'm not sure why that is:
https://github.com/tensorflow/tensorflow/blob/ced4fddc0400fedce8ba15fb9a3e6e765ddda757/tensorflow/core/distributed_runtime/rpc/eager/grpc_eager_service.h#L43
https://github.com/tensorflow/tensorflow/blob/ced4fddc0400fedce8ba15fb9a3e6e765ddda757/tensorflow/core/distributed_runtime/rpc/grpc_worker_service_impl.h#L98

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-4562 about the composite byte buffer optimization

**Comment by wesm** [comment]
I'm sure you could ask TF about it and find the right person to talk to. 

You might want to try benchmarking with two of the nodes in the UL Nashville network rather than localhost-to-localhost. They are connected with gigabit ethernet only (I'm upgrading the network switches but not sure I can get 10-gigabit working, too many factors out of my control) so it would be interesting to see to what extent we are saturating the network bandwidth

**Comment by wesm** [comment]
@pitrou regarding padding, I could imagine an application that hashes IPC payloads for some purpose (caching?) having an issue with non-zero padding. It seems like an esoteric concern; I'm not sure it is worth paying the price, or at least having the "sanitize padding" option as something that's disabled by default

**Comment by pitrou** [comment]
Perhaps the price isn't that large either. Ideally the batches are large so adding a 8-bytes zero-initialized slice may be cheap...

File: cpp/src/arrow/flight/protocol.cc (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/flight/protocol.cc]: done

**Comment by wesm** [comment]
OK to merge this with a passing build? This could benefit from an IWYU pass but I might like to do that for all of Flight to clean up all the includes and forward-declarations

**Comment by wesm** [comment]
I should rename protocol.h to protocol-internal.h so the header is not installed.

**Review by pitrou — COMMENTED** [review]
LGTM, one comment.

File: cpp/src/arrow/flight/serialization-internal.cc [hunk 3]
```diff
@@ -32,6 +32,214 @@ bool ReadBytesZeroCopy(const std::shared_ptr<arrow::Buffer>& source_data,
   return input->Skip(static_cast<int>(length));
 }
 
+using google::protobuf::io::CodedInputStream;
+using google::protobuf::io::CodedOutputStream;
+
+bool ReadBytesZeroCopy(const std::shared_ptr<arrow::Buffer>& source_data,
```
Comment (pitrou) [inline cpp/src/arrow/flight/serialization-internal.cc]: Oh, and you needn't redeclare this (already defined above).

File: cpp/src/arrow/flight/customize_protobuf.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <limits>
+#include <memory>
+
+#include "grpcpp/impl/codegen/config_protobuf.h"
+#undef GRPC_OPEN_SOURCE_PROTO
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/flight/customize_protobuf.h]: done

File: cpp/src/arrow/flight/serialization-internal.cc (hunk 3 above)
Comment (wesm) [reply cpp/src/arrow/flight/serialization-internal.cc]: done

**Comment by wesm** [comment]
OK I'll wait for the build to run again and then merge

**Comment by wesm** [comment]
Merging. A couple of builds experienced transient failures

**Pull request closed without merging**